db/hybrid_logical_vector_test.go

```diff
@@ -0,0 +1,187 @@
+// Copyright 2023-Present Couchbase, Inc.
+//
+// Use of this software is governed by the Business Source License included
+// in the file licenses/BSL-Couchbase.txt.  As of the Change Date specified
+// in that file, in accordance with the Business Source License, use of this
+// software will be governed by the Apache License, Version 2.0, included in
+// the file licenses/APL2.txt.
+
+package db
+
+import (
+	"reflect"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+// TestInternalHLVFunctions:
+//   - Tests internal api methods on the HLV work as expected
+//   - Tests methods GetCurrentVersion, AddVersion and Remove
+func TestInternalHLVFunctions(t *testing.T) {
+	mv := make(map[string]uint64)
+	pv := make(map[string]uint64)
+	currSourceId := "s_5pRi8Piv1yLcLJ1iVNJIsA"
+	const currVersion = 12345678
+
+	mv["s_NqiIe0LekFPLeX4JvTO6Iw"] = 345454
+	pv["s_YZvBpEaztom9z5V/hDoeIw"] = 64463204720
+
+	hlv := HybridLogicalVector{
+		CurrentVersionCAS: currVersion,
+		SourceID:          currSourceId,
+		Version:           currVersion,
+		MergeVersions:     mv,
+		PreviousVersions:  pv,
+	}
+
+	const newCAS = 123456789
+	const newSource = "s_testsource"
+	currVersionVector := CurrentVersionVector{
+		VersionCAS: 123456789,
+		SourceID:   newSource,
+	}
+
+	pv[currSourceId] = currVersion
+
+	// Get current version vector, sourceID and CAS pair
+	source, version := hlv.GetCurrentVersion()
+	assert.Equal(t, currSourceId, source)
+	assert.Equal(t, uint64(currVersion), version)
+
+	// Add a new version vector pair to the HLV structure and assert that it moves the current version vector pair to the previous versions section
+	hlv.AddVersion(currVersionVector)
+	assert.Equal(t, uint64(newCAS), hlv.Version)
+	assert.Equal(t, newSource, hlv.SourceID)
+	assert.True(t, reflect.DeepEqual(hlv.PreviousVersions, pv))
+
+	// Remove a sourceID CAS pair from previous versions section of the HLV structure (for compaction)
+	hlv.Remove(currSourceId)
+	delete(pv, currSourceId)
+	assert.True(t, reflect.DeepEqual(hlv.PreviousVersions, pv))
+}
+
+// TestConflictDetectionDominating:
+//   - Tests two cases where one HLV's is said to be 'dominating' over another and thus not in conflict
+//   - Test case 1: where sourceID is the same between HLV's but the in memory representation has higher version CAS
+//   - Test case 2: where sourceID is different and the in memory HLV has higher CAS than other HLV but the other HLV's
+//     sourceID:CAS pair is present in the in memory HLV's previous versions
+//   - Assert that both scenarios returns false from IsInConflict method, as we have a HLV that is dominating in each case
+func TestConflictDetectionDominating(t *testing.T) {
```

Comment:
I think tests for conflicts should define HLVs with multiple versions, instead of just a single version per HLV.  (the latter is a valid case, but only one of many).  For testing it would be useful to define a human readable shorthand for the HLV that could be parsed into the HLV struct.  You could come up with a proposal, I'm thinking something like:
   [ver@src, ver@src, ver@src, ver@src]
This would assume the 0th entry in the list is current version, and everything else is a previous version.  Adding some type of annotation for merge versions would be needed, maybe m_ver@src?